commit 4cf04c6a45328d0d0cd083be12426d52976db1cd
Author: root <root@took23.ics.muni.cz>
Date:   Wed Oct 5 15:01:54 2016 +0200

    meta-initrefreshcc.patch
    
    init.d/rsyslog refreshcc
    systemd service refreshcc
    oprava bugu v kinitu pokud existuje defaultni ccache
    https://rt4.cesnet.cz/rt/Ticket/Display.html?id=126655

diff --git a/debian/rsyslog.init b/debian/rsyslog.init
index 0389df6..57313e2 100644
--- a/debian/rsyslog.init
+++ b/debian/rsyslog.init
@@ -14,6 +14,7 @@
 
 #
 # Author: Michael Biebl <biebl@debian.org>
+# bodik@civ.zcu.cz
 #
 
 # PATH should only include /usr/* if it runs after the mountnfs.sh script
@@ -38,6 +39,13 @@ SCRIPTNAME=/etc/init.d/$NAME
 
 do_start()
 {
+	
+#	export RSYSLOG_DEBUGLOG="/var/log/rsyslogd"
+#	export RSYSLOG_DEBUG="Debug NoStdOut"
+	do_refreshcc
+	ulimit -c unlimited
+
+
 	# Return
 	#   0 if daemon has been started
 	#   1 if daemon was already running
@@ -59,6 +67,15 @@ do_stop()
 #
 do_rotate() {
 	start-stop-daemon --stop --signal HUP --quiet --pidfile $PIDFILE --exec $DAEMON
+	RET=$?
+
+	pidof rsyslogd 1>/dev/null 2>&1
+	if [ $? -eq 1 ]; then
+		#nebezi
+		rm -f $RSYSLOG_CCFILE
+	fi
+
+	return $RET
 }
 
 create_xconsole() {
@@ -80,6 +97,46 @@ sendsigs_omit() {
 	ln -sf $PIDFILE $OMITDIR/rsyslog
 }
 
+RSYSLOG_CCFILE="/tmp/krb5cc_rsyslog"
+
+do_refreshcc()
+{
+	export KRB5CCNAME="FILE:$RSYSLOG_CCFILE"
+
+	if [ ! -f /etc/krb5.keytab ]; then
+        	logger -p syslog.info -t "rsyslog.init[$$]" "refreshcc skipped, missing keytab"
+		return
+	fi
+
+	#MIT
+	klist --version 1>/dev/null 2>&1
+	if [ $? -eq 0 ]; then
+		#Heimdal
+		PRINC=`ktutil -k /etc/krb5.keytab list | grep "host/" | head -1 | awk '{print $3}'`
+		OUT=`kinit --no-afslog -c $KRB5CCNAME -k $PRINC 2>&1`
+	else
+		#MIT ;)
+		PRINC=`klist -k | grep "host/" | head -1 | awk '{print $2}'`
+		OUT=`kinit -c $KRB5CCNAME -k $PRINC 2>&1`
+	fi
+	RET=$?
+
+
+	if [ $RET -ne 0 ]; then 
+		#echo "DEBUG: kinit ret $RET"
+		logger -s -p syslog.error -t "rsyslog.init[$$]" "refreshcc failed: $OUT"
+		rm -f $RSYSLOG_CCFILE
+		RJOBS=`grep -shc refreshcc /var/spool/cron/atjobs/* | awk 'BEGIN{sum=0}{sum+=$1}END{print sum}'`
+		if [ $RJOBS -le 1 ]; then #pokud bezim a odfailoval sem stejne se chci naplanovat
+			logger -s -p syslog.info -t "rsyslog.init[$$]" "refreshcc scheduling fast refresh"
+		        echo "$0 refreshcc &>/dev/null" | at "now + 19 minutes"
+		fi
+	else 
+		logger -p syslog.info -t "rsyslog.init[$$]" "refreshcc ok"
+	fi
+}
+
+
 case "$1" in
   start)
 	log_daemon_msg "Starting $DESC" "$RSYSLOGD"
@@ -117,6 +174,9 @@ case "$1" in
   status)
 	status_of_proc -p $PIDFILE $DAEMON $RSYSLOGD && exit 0 || exit $?
 	;;
+  refreshcc)
+	do_refreshcc
+	;;
   *)
 	echo "Usage: $SCRIPTNAME {start|stop|rotate|restart|force-reload|status}" >&2
 	exit 3
diff --git a/rsyslog.service.in b/rsyslog.service.in
index cb629ee..8f21e9b 100644
--- a/rsyslog.service.in
+++ b/rsyslog.service.in
@@ -6,6 +6,9 @@ Documentation=http://www.rsyslog.com/doc/
 
 [Service]
 Type=notify
+Environment="KRB5CCNAME=FILE:/tmp/krb5cc_rsyslog"
+ExecStartPre=/bin/sh -c "/etc/init.d/rsyslog refreshcc"
+ExecReload=/bin/sh -c "/etc/init.d/rsyslog refreshcc"
 ExecStart=@sbindir@/rsyslogd -n
 StandardOutput=null
 Restart=on-failure
